client/client.go

```diff
@@ -22,57 +22,318 @@ limitations under the License.
+	inspectMutex sync.Mutex
 }
 
-type dockerClient struct{}
+type deviceInfo struct {
+	device string
+	major  string
+	minor  string
+}
+
+func init() {
+	getDockerVersion := func() []int {
+		dc := NewDockerClient()
+		version, err := dc.Version()
+		if err != nil {
+			panic(err)
+		}
+		return version
+	}
+	config.DockerVersion = getDockerVersion()
+	fmt.Fprintf(os.Stderr, "Docker runtime version: %v \n", config.DockerVersion)
+}
 
 func NewDockerClient() *dockerClient {
-	return new(dockerClient)
+	client, err := docker.NewClient(endpoint)
+	if err != nil {
+		panic(err)
+	}
+
+	return &dockerClient{cl: client,
+		inspectCache: map[string]*docker.Container{},
+	}
 }
 
 func (dc *dockerClient) FindCgroupMountpoint(subsystem string) (string, error) {
 	return cgroups.FindCgroupMountpoint(subsystem)
 }
 
-func (dc *dockerClient) ListContainers() ([]ContainerInfo, error) {
-	cl, err := docker.NewClient(endpoint)
+func (dc *dockerClient) InspectContainer(id string) (*docker.Container, error) {
+	dc.inspectMutex.Lock()
+	defer dc.inspectMutex.Unlock()
+	if info, haveInfo := dc.inspectCache[id]; !haveInfo {
+		if info, haveInfo = dc.inspectCache[id]; haveInfo {
+			// double-locking is fine here, as side effects are harmless
+			return info, nil
+		}
+		var err error
+		if info, err = dc.cl.InspectContainer(id); err != nil {
+			return nil, err
+		} else {
+			dc.inspectCache[id] = info
+			return info, nil
+		}
+	} else {
+		return info, nil
+	}
+}
+
+// Version returns version of docker engine
+func (dc *dockerClient) Version() (version []int, _ error) {
+	version = []int{0, 0}
+	env, err := dc.cl.Version()
 	if err != nil {
-		fmt.Println("[ERROR] Could not create docker client!")
-		return []ContainerInfo{}, err
+		return version, err
 	}
+	parseInt := func(str string, defVal int) int {
+		if val, err := strconv.ParseInt(str, 10, 64); err != nil {
+			return defVal
+		} else {
+			return int(val)
+		}
+	}
+	for _, kv := range *env {
+		kvs := strings.Split(kv, "=")
+		if kvs[0] != dockerVersionKey {
+			continue
+		}
+		versionSplit := strings.Split(kvs[1], ".")
+		version := []int{parseInt(versionSplit[0], 0), parseInt(versionSplit[1], 0)}
+		return version, nil
+	}
+	return version, nil
+}
+
+func (dc *dockerClient) ListContainersAsMap() (map[string]docker.APIContainers, error) {
+	containers := make(map[string]docker.APIContainers)
+
+	containerList, err := dc.cl.ListContainers(docker.ListContainersOptions{})
 
-	containerList, err := cl.ListContainers(docker.ListContainersOptions{})
 	if err != nil {
-		fmt.Println("[ERROR] Could not retrieve list of running containers!")
-		return []ContainerInfo{}, err
+		return nil, err
 	}
 
-	containers := []ContainerInfo{}
 	for _, cont := range containerList {
-		cinfo := ContainerInfo{
-			Id:      cont.ID,
-			Status:  cont.Status,
-			Created: cont.Created,
-			Image:   cont.Image}
-		containers = append(containers, cinfo)
+		shortID, _ := GetShortId(cont.ID)
```

Comment:
ok, handling added -> but if cont.ID is invalid that will mean that docker API returns invalid values and such error will be handled earlier.